xkafka/consumer.go

```diff
@@ -0,0 +1,162 @@
+package xkafka
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/sourcegraph/conc/stream"
+
+	"github.com/gojekfarm/xtools/xkafka/internal"
+)
+
+// ConsumerFunc allows the user to customise confluent-kafka-go/kafka.Consumer.
+type ConsumerFunc func(cfg *kafka.ConfigMap) (internal.ConsumerClient, error)
+
+func (cf ConsumerFunc) apply(o *options) { o.consumerFn = cf }
+
+// DefaultConsumerFunc is the default ConsumerFunc that initialises
+// a new confluent-kafka-go/kafka.Consumer.
+func DefaultConsumerFunc(cfg *kafka.ConfigMap) (internal.ConsumerClient, error) {
+	return kafka.NewConsumer(cfg)
+}
+
+// Consumer manages the consumption of messages from kafka topics
+// and the processing of those messages.
+type Consumer struct {
+	name        string
+	kafka       internal.ConsumerClient
+	handler     Handler
+	middlewares []middleware
+	config      options
+}
+
+// NewConsumer creates a new Consumer instance.
+func NewConsumer(name string, opts ...Option) (*Consumer, error) {
+	cfg := defaultConsumerOptions()
+
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	_ = cfg.configMap.SetKey("bootstrap.servers", strings.Join(cfg.brokers, ","))
+	_ = cfg.configMap.SetKey("group.id", name)
+
+	consumer, err := cfg.consumerFn(&cfg.configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Consumer{
+		name:   name,
+		config: cfg,
+		kafka:  consumer,
+	}, nil
+}
+
+// GetMetadata returns the metadata for the consumer.
+func (c *Consumer) GetMetadata() (*kafka.Metadata, error) {
+	return c.kafka.GetMetadata(nil, false, int(c.config.metadataTimeout.Milliseconds()))
+}
+
+// Use appends a MiddlewareFunc to the chain.
+// Middleware can be used to intercept or otherwise modify, process or skip messages.
+// They are executed in the order that they are applied to the Consumer.
+func (c *Consumer) Use(mwf ...MiddlewareFunc) {
+	for _, fn := range mwf {
+		c.middlewares = append(c.middlewares, fn)
+	}
+}
+
+// WithHandler sets the message handler for the consumer.
+// Any previously set handlers are overwritten.
+func (c *Consumer) WithHandler(handler Handler) *Consumer {
```

Comment:
This can be taken as a required argument in `NewConsumer` or as an `Option`?

Reply:
I don't think we need a separate function.

Reply:
made handler a required param